packages/salesforce-adapter/src/filters/layouts.ts

```diff
@@ -77,12 +78,68 @@ const fixLayoutPath = (
   layout.path = [...objectPath.slice(0, -1), layout.elemID.typeName, layout.elemID.name]
 }
 
+type LayoutItem = {
+  field: string | ReferenceExpression
+}
+
+type LayoutColumn = {
+  layoutItems: LayoutItem[]
+}
+
+type LayoutSection = {
+  layoutColumns: LayoutColumn[]
+}
+
+const fixLayoutItemField = (
+  layoutItem: LayoutItem, layoutObj: ObjectType, layoutObjFields: Field[]
+): void => {
+  const findField = (fieldName: string, fields: Field[]): Field | undefined =>
+    fields.find(field => field.name === fieldName)
+  const reference = findField(layoutItem.field as string, layoutObjFields)
+  if (!reference) {
+    log.debug(`Could not find field ${layoutItem.field} for layout ${layoutObj.elemID.name}`)
+    return
+  }
+
+  layoutItem.field = new ReferenceExpression(
+    reference.elemID
+  )
+}
+
+const fixLayoutColumnItems = (
+  layout: InstanceElement,
+  layoutObj: ObjectType,
+  layoutObjFields: Field[]
+): void => {
+  const fixToArray = (obj: unknown): unknown => (_.isArray(obj) ? obj : [obj])
+
+  if (!_.has(layout.value, 'layoutSections')) return
+  const layoutSections = fixToArray(layout.value.layoutSections) as LayoutSection[]
+
+  layoutSections.forEach((layoutSection: LayoutSection) => {
+    if (!_.has(layoutSection, 'layoutColumns')) return
+    const layoutColumns = fixToArray(layoutSection.layoutColumns) as LayoutColumn[]
+
+    layoutColumns.forEach((layoutColumn: LayoutColumn) => {
+      if (!_.has(layoutColumn, 'layoutItems')) return
+      const layoutItems = fixToArray(layoutColumn.layoutItems) as LayoutItem[]
+
+      layoutItems.forEach((layoutItem: LayoutItem) => fixLayoutItemField(
+        layoutItem, layoutObj, layoutObjFields
+      ))
+    })
+  })
+}
```

Comment:
there is a mismatch between the typescript type and the actual value which makes this code strange.
the typescript type you made says `layoutColumns` is not optional, so this check doesn't make sense but in fact (according to this code) `layoutColumns` is not always there and is actually optional.

with the proper types this can be much more concise
```suggestion
  const layoutSections = makeArray(layout.value.layoutSections) as LayoutSection[]
  const layoutItems = _(layoutSections)
    .map(section => makeArray(section.layoutColumns))
    .flatten()
    .map(column => makeArray(column.layoutItems))
    .flatten()
  layoutItems.forEach(item => fixLayoutItemField(item, ...))
}
```

Reply:
Fixed but becasue i still need fixToArray. There's a little more casting than in your example.

Reply:
if you used `makeArray` it should preserve the types so you wouldn't have to cast so much (I am pretty sure)

Reply:
Your right. Sorry for the mix up